cpp/src/arrow/compute/kernels/scalar_arithmetic_test.cc

```diff
@@ -37,68 +37,275 @@
+    // Addition
+    auto left = this->MakeInputArray({max, min});
+    auto right = this->MakeInputArray({1, -1});
+    auto expected = this->MakeOutputArray({min, max});
+    this->AssertBinop(arrow::compute::Add, left, right, expected);
+  }
+  {
+    // Subtraction
+    auto left = this->MakeInputArray({min, max});
+    auto right = this->MakeInputArray({1, -1});
+    auto expected = this->MakeOutputArray({max, min});
+    this->AssertBinop(arrow::compute::Subtract, left, right, expected);
+  }
+  {
+    // Multiplication
+    auto left = this->MakeInputArray({min, max});
+    auto right = this->MakeInputArray({-1, 2});
+    auto expected = this->MakeOutputArray({min, -2});
+    this->AssertBinop(arrow::compute::Multiply, left, right, expected);
+  }
+}
+
+TYPED_TEST(TestBinaryArithmeticsSigned, Sub) {
+  this->AssertBinop(arrow::compute::Subtract, "[0, 1, 2, 3, 4, 5, 6]",
+                    "[1, 2, 3, 4, 5, 6, 7]", "[-1, -1, -1, -1, -1, -1, -1]");
+
+  this->AssertBinop(arrow::compute::Subtract, "[0, 0, 0, 0, 0, 0, 0]",
+                    "[6, 5, 4, 3, 2, 1, 0]", "[-6, -5, -4, -3, -2, -1, 0]");
+
+  this->AssertBinop(arrow::compute::Subtract, "[10, 12, 4, 50, 50, 32, 11]",
+                    "[2, 0, 6, 1, 5, 3, 4]", "[8, 12, -2, 49, 45, 29, 7]");
+
+  this->AssertBinop(arrow::compute::Subtract, "[null, 1, 3, null, 2, 5]",
+                    "[1, 4, 2, 5, 0, 3]", "[null, -3, 1, null, 2, 2]");
+}
+
+TYPED_TEST(TestBinaryArithmeticsSigned, Mul) {
+  this->AssertBinop(arrow::compute::Multiply, "[-10, 12, 4, 50, -5, 32, 11]",
+                    "[-2, 0, -6, 1, 5, 3, 4]", "[20, 0, -24, 50, -25, 96, 44]");
+}
+
+TYPED_TEST(TestBinaryArithmeticsFloating, Add) {
+  this->AssertBinop(arrow::compute::Add, "[]", "[]", "[]");
+
+  this->AssertBinop(arrow::compute::Add, "[3.4, 2.6, 6.3]", "[1, 0, 2]",
+                    "[4.4, 2.6, 8.3]");
+
+  this->AssertBinop(arrow::compute::Add, "[1.1, 2.4, 3.5, 4.3, 5.1, 6.8, 7.3]",
+                    "[0, 1, 2, 3, 4, 5, 6]", "[1.1, 3.4, 5.5, 7.3, 9.1, 11.8, 13.3]");
+
+  this->AssertBinop(arrow::compute::Add, "[7, 6, 5, 4, 3, 2, 1]", "[6, 5, 4, 3, 2, 1, 0]",
+                    "[13, 11, 9, 7, 5, 3, 1]");
+
+  this->AssertBinop(arrow::compute::Add, "[10.4, 12, 4.2, 50, 50.3, 32, 11]",
+                    "[2, 0, 6, 1, 5, 3, 4]", "[12.4, 12, 10.2, 51, 55.3, 35, 15]");
+
+  this->AssertBinop(arrow::compute::Add, "[null, 1, 3.3, null, 2, 5.3]",
+                    "[1, 4, 2, 5, 0, 3]", "[null, 5, 5.3, null, 2, 8.3]");
+
+  this->AssertBinop(arrow::compute::Add, 1.1, "[null, 1, 3.3, null, 2, 5.3]",
+                    "[null, 2.1, 4.4, null, 3.1, 6.4]");
+}
+
+TYPED_TEST(TestBinaryArithmeticsFloating, Sub) {
+  this->AssertBinop(arrow::compute::Subtract, "[]", "[]", "[]");
+
+  this->AssertBinop(arrow::compute::Subtract, "[3.4, 2.6, 6.3]", "[1, 0, 2]",
+                    "[2.4, 2.6, 4.3]");
 
-  this->AssertAdd("[3, 2, 6]", "[1, 0, 2]", "[4, 2, 8]");
+  // this->AssertBinop(arrow::compute::Subtract,
```

Comment:
The array equality check fails despite that after printing out both sides the values are the same. I'm not sure why, perhaps a float precision problem?